### ISSUE TITLE
Using the between_time argument

### ISSUE THREAD

**Issue by markxbaker** [edtechre/pybroker#167]
Hi Ed,

Is the between_time argument supposed to only use the bars you set to  enter _and_ exit?

For example say I'm testing an sma cross on 30min timeframe as a buy set with between_time 9:30am and 12:00am - this triggers the buy between these times if the cross happens which is as expected.

But with hold bars set to say 4 or 6 the exit should be same day but outside of these times -  but all of the exits are also between these times the next day so doesn't look like hold bars is working correctly here.

I'd have thought the logic would be for entries to happen between these times but exits to trigger as and when they occur, e.g hold bars or other condition triggers.

Can workaround by creating a time column  as part of building a data source but would be cleaner if handled in the package and could use API data sources this way also.

Also if there was some way to set a of flag for conditions to sit outside these times also, for example another cross to exit should happen when it occurs on any of the bars.

Something liek this

# Check for entry condition
    if not long_pos and cross_up:# and natr[-1] < natr_moving_average[-1]:

          target_shares = 100

          ctx.buy_shares = target_shares
          ctx.buy_fill_price = PriceType.OPEN
          ctx.hold_bars = 12
          ctx.hold_bars_exit = ALL_HOURS
          ctx.stop_loss = ATR

    elif long_pos and cross_down and EXIT_FLAG=ALL_AVAILABLE_HOURS:
      ctx.sell_all_shares()
      ctx.sell_fill_price = pybroker.common.PriceType.OPEN

# Set up strategy and run backtest
strategy = Strategy(data_source_M30, start_date, end_date, config)
strategy.add_execution(exec_fn, tickers, indicators=list(indicators.values()))
strategy.set_slippage_model(RandomSlippageModel(min_pct=1.0, max_pct=5.0))

# Run backtest
result = strategy.backtest(warmup=52, calc_bootstrap=True between_time=('9:30', '11:00'))

Thanks!

Mark

**Comment by clivinn-shla81092** [comment]
Hello, we tried to solve the issue.

This is what we did:

This change adds new parameters to the Strategy class to allow for separate entry and exit time ranges, as well as flags for conditions that can occur outside these times. It also modifies the backtest and walkforward methods to use these new parameters.

You can review changes in this commit: https://github.com/clivinn-shla81092/edtechre-pybroker-167/commit/05133a1beccc3e44f528cc1963e009a0ea38261b.


> [!CAUTION]
> Disclaimer: The concept of solution was created by AI and you should never copy paste this code before you check the correctness of generated code. Solution might not be complete, you should use this code as an inspiration only.

---

Latta AI seeks to solve problems in open source projects as part of its mission to support developers around the world. Learn more about our mission at https://latta.ai/ourmission . If you no longer want Latta AI to attempt solving issues on your repository, you can block this account.

**Comment by markxbaker** [comment]
Wow that was fast! How would I then use it in my code?

Thanks,

Mark

**Comment by markxbaker** [comment]
Added to custom data source works the way I expected.
